File: Colony.py
import numpy as np
import Ant as at
import random

class Colony(object):
	"""
	Ant container class, handles ant generation.
	
	@param config: Configuration parameters
	
	@field size: Tile Size in pixels
	@field screen: Size of the screen
	@field default: Default Ruleset
	@field sym: Symbols used to describe a ruleset
	@field length: Length of each ant ruleset
	@field dir: Dictionary containing unitary vectors for each direction
	@field ants: List containing the Ant objects
	"""
	def __init__(self, config, tile):
		
		self.screen = (config.getint('Display', 'WIDTH'), 
						config.getint('Display', 'HEIGHT'))
		self.default = config.get('Ruleset', 'DEFAULT')
				
		self.sym = list(config.get('Ruleset', 'SYMBOLS'))
		self.length = config.getint('Ruleset', 'LENGTH')
		self.save = config.getboolean('Ruleset', 'SAVE')
		
		self.Ttype = tile
		
		self.ants = []
		
		if self.default == 'RANDOM':
			self.default = self.random_ruleset()
		
	def load_ants(self, path):
		"""
		Loads ant data from a txt file
		
		@param: path, File path to be used
		"""
		with open(path, "r") as file:
			for line in file:
				data = line.strip("\n").split("\t")
				
				if len(data) == 4:	#Ruleset exists
					if data[3] == 'RANDOM':
						rule = self.random_ruleset()
					else:
						assert len(data[3]) == self.length, "Ruleset Error: Inconsistent ruleset size"
						rule = data[3]
				else:
					rule = self.default
				
				print(rule)
				if data[0] == "R":
					data[0] = random.randint(1,self.screen[0]-1)
				if data[1] == "R":
					data[1] = random.randint(1,self.screen[1]-1)
				if data[2] == "R":
					data[2] = random.randint(0,len(self.Ttype.directions.keys())-1)
				
				bug = at.Ant(int(data[0]) * self.Ttype.size,
							int(data[1]) * self.Ttype.size,
							int(data[2]), rule, self.Ttype)
							
				self.ants.append(bug)
				
		if self.save:
			self.save_ants("save/ants.txt")
	
	def save_ants(self, path):
		"""
		Saves data of generated ants in a backup txt file
		
		@param path: Location of the backup
		"""
		f = open(path,"w")
		dir = self.Ttype.directions
		size = self.Ttype.size
		for ant in self.ants:
			for k in dir.keys():
				if k== ant.dir:
					face = str(k)
					break
			f.write(str(ant.pos[0]//size)+"\t"+str(ant.pos[1]//size)+"\t"+face+"\t"+ant.ruleset+"\n")
		f.close()
		
		
		
	def random_ruleset(self):
		"""
		Generates a random set of instructions from the given specifications
		
		@returns:	String containing the new ruleset
		
		"""
		rand_rule = ""
		for i in range(self.length):
			ch = random.randint(0, len(self.sym)-1)
			rand_rule += self.sym[ch]
		
		return rand_rule

		

File: C++/Colony.cpp
/***
 * @file Colony.cpp
 * Ant colony implementation
 */
// Copyright 2019 by Julio Albornoz <gatalegath@protonmail.com>
// The License.txt file describes the conditions under which this software may be distributed.

#include "Colony.h"
#include <sstream>
#include <time.h>

Colony::Colony(Config* config, Tile* tile){
	screen[0] = config->DWidth;
	screen[1] = config->DHeight;
	
	def_rule = config->def_rule;
	sym = config->symbols;
	length = config->length;
	save = config->Rsave;
	
	Ttype = tile;
	
	if (def_rule.compare("RANDOM") == 0){
		def_rule = random_ruleset();
		}
		
	load_ants();
	}
	
Colony::~Colony(){
	delete ants;
	}
	
void Colony::load_ants(std::string path){
	/***
	* Loads ant data from a txt file
	* 	
	* @param path: File path to be used
	*/ 
	std::ifstream file(path);	//File content
	std::string line;
	
	srand (time(NULL));
	std::vector<Ant>* insects = new std::vector<Ant>; 
	while (std::getline(file, line)){	//Read entire line
		std::stringstream linestream(line);
		
		std::string buffer;
		int X, Y, dir;		// Move parameters to an Ant factory
		
		//X position
		std::getline(linestream, buffer, '\t');		//Refactor this block
		if (buffer.compare("R") == 0){
			X = rand() % (screen[0] - 1) + 1;
			}
		else{
			X = std::stoi(buffer);
			}
			
		buffer.clear();
		
		//Y position
		std::getline(linestream, buffer, '\t');		//Refactor this block
		if (buffer.compare("R") == 0){
			Y = rand() % (screen[1] - 1) + 1;
			}
		else{
			Y = std::stoi(buffer);
			}
			
		buffer.clear();
		
		//Ant orientation
		std::getline(linestream, buffer, '\t');		//Refactor this block
		if (buffer.compare("R") == 0){
			int max = Ttype->directions.size();
			dir = rand() % max;
			}
		else{
			dir = std::stoi(buffer);
			}
		
		buffer.clear();
		
		//Ant ruleset
		if (linestream.eof()){			//Uses default ruleset
			buffer = def_rule;
			}
		
		std::getline(linestream, buffer);	//4th column
		if (buffer.compare("RANDOM") == 0){
			buffer = random_ruleset();
			}
		
		// Initializes Ant
		Ant bug = Ant(X * Ttype->X, Y * Ttype->Y, dir, buffer, Ttype);
		insects->push_back(bug);
		}
		
	ants = insects;
	file.close();
	
	if (save){
		save_ants("save/ants.txt");
		}
	}

void Colony::save_ants(std::string path){
	/***
	* Saves data of generated ants in a backup txt file
	* 	
	* @param path: Location of the backu
	*/
	std::ofstream file(path);	//File content
	std::vector<Ant> bugs = *ants;
	
	for (unsigned int i = 0; i < bugs.size(); i++){
		Ant ant = bugs[i];
		file << ant.pos[0]/Ttype->X << "\t" << ant.pos[1]/Ttype->Y << "\t" << ant.dir << "\t" << ant.ruleset << "\n";	
		}
	file.close();
	}

std::string Colony::random_ruleset(){
	/***
	* Generates a random set of instructions from the given specifications
	* 
	* @return	String containing the new ruleset
	* 
	*/
	srand (time(NULL));
	std::string Rrule;
	for (unsigned int i = 0; i < length; i++){
		int ch = rand() % sym.size();
		Rrule += sym[ch];
		}
	return Rrule;
	}
	
